### ISSUE TITLE
RewritingVisitor does not visit some dictionaries

### ISSUE THREAD

**Issue by lgolding** [Microsoft/jschema#37]
For example, it visits neither the `MessageStrings` nor the `Rules` dictionary-valued property of `Resources`:
```
        public virtual Resources VisitResources(Resources node)
        {
            if (node != null)
            {
            }

            return node;
        }
```

and it does not visit `Run.Graphs` or `Result.Graphs`.

It _does_ visit `Run.Files`, so that will be a good clue.

@michaelcfanning FYI

**Comment by lgolding** [comment]
- The dictionaries it doesn't visit are those with a non-default value type.
- The `DictionaryHint` unit tests do not examine the rewriting visitor at all.

@michaelcfanning FYI, this _is_ a bug in the `RewritingVisitorGenerator` class, and the unit tests for `DictionaryHint` have absolutely no coverage on this. I'd love to have blamed this on someone else, but I looked at the history and I wrote every line of this ;-)

**Comment by lgolding** [comment]
@EasyRhinoMSFT @michaelcfanning
 
I investigated this and it is _not_ a bug in the JSchemaToDotNet code generator. The code works as intended and I am closing this bug `resolved-by-design`. This completes my work on this issue.

The proximate cause of the bug in the sarif-sdk (the bug being that certain dictionaries were not visited) is that I misused the arguments to `DictionaryHint`. The root cause was that the intended usage of those arguments is not documented (in fact, nothing about the usage of the code generator is documented).

The code generator will only generate code to visit the _elements_ of a dictionary if those elements are of a type defined by the schema. This makes sense, because only for schema-defined types does the code generator know how to visit the _properties_ of those dictionary elements.

If you do _not_ supply the `valueType` argument to the `DictionaryHint`, the code generator examines the schema to determine the element type. If the schema says that the element type is itself defined in the schema (for example, if the element type is `"$ref": "#/definitions/graph"`, then and only then does the code generator decide that the element is of schema-defined type, and then and only then does it generate code to visit the dictionary.

On the other hand, if you _do_ specify the `valueType` argument to the `DictionaryHint`, the code generator assumes that you did this is because the value type is _not_ defined in the schema. Therefore it does not generate code to visit the dictionary.

My bug in the sarif-sdk was that for the three dictionary-valued properties `run.graphs`, `result.graphs`, and `resources.rules`, whose element types _are_ defined in the schema, I gratuitously specified the `valueType` argument to the `DictionaryHint`. That is now fixed.
